Sources/SwiftFusion/Nonlinear/NonlinearFactorGraph.swift

```diff
@@ -0,0 +1,114 @@
+import TensorFlow
+
+public struct NonlinearFactorGraph {
+  var factor1s: [Factor1] = []
```

Comment:
I think all we need is be polymorphic about factor type. But maybe this "old thinking" to have a abstract base class? What the new way? Protocol?

Reply:
I think we can have a `Factor` protocol that defines a loss function, and a `DifferentiableFactor` with both loss and  derivative.

Reply:
Ideally we would not use classes because they do not interact with AD very well:
 * the AD implementation for classes isn't very robust or fleshed out yet, and
 * even if the implementation was robust, AD is probably always going to behave in weird unexpected ways in the presence of classes because classes have [reference semantics](https://developer.apple.com/swift/blog/?id=10) and it's not obvious what it means to differentiate functions involving references to things.

Also, structs are just generally more encouraged than classes in Swift, so it's good to use them when possible.

We should indeed be able to do this with some kind of `Factor` protocol and structs that conform to that protocol. (Why a separate `DifferentiableFactor` protocol? Can we just make `Factor` have all the necessary stuff for differentiation?) Once we have a `Factor` protocol, we can store an array `[Factor]` in the factor graph. (The type `Factor` is called an "existential" and it can store values of any type that conforms to `Factor`).

There is one difficulty with this approach. The obvious `Factor` protocol would be something like this:
```
protocol Factor {
  associatedtype InputType: Differentiable

  @differentiable
  func error(_ input: InputType) -> Tensor<Double>
}
```

Unfortunately, Swift does not support existentials for protocols with `associatedtype`s, so we can't store an array of those in the factor graph.

We can work around this by defining something like:

```
protocol Factor {
  func error(anyInput: Any) -> Tensor<Double>
  func errorPullback(anyInput: Any, _ v: Tensor<Double>) -> Tensor<Double>
}
```

This has no associated types, so we can use its existential.

Then we can define helper protocols so that users defining factors do not need to do all the casting themselves:

```
protocol TypedFactor1 {
  associatedtype InputType: Differentiable
  @differentiable func error(_ input: InputType) -> Tensor<Float>
}

extension TypedFactor1 {
  func error(anyInput: Any) -> Tensor<Double> {
    let input = anyInput as! InputType
    return error(input)
  }

  func errorPullback(anyInput: Any) -> (Tensor<Double>) -> Tensor<Double> {
    let input = anyInput as! InputType
    return pullback(at: input, in: { error($0) })
  }
}

struct PriorFactor: TypedFactor1, Factor {
  let prior: Pose2

  @differentiable
  func error(_ input: Pose2) -> Tensor<Float> {
    between(input, prior).tensor
  }
}
```

I think this is probably the part of the API that I can be most helpful on, so I'm going to play with this and see if I can get a working prototype to show you before we have a call.

Reply:
Yes, that sounds great!
Note there is another, more ad-hoc way in GTSAM to create factors which is "AddExpressionFactor", where you give an expression (created in a compile time DSL), a measurement, and a noise model. The key feature there is to have "variable expressions" that refer to the Values through a key - very much  like tensorflow "Variables"  in fact. Obvious question: can we hijack the machinery in swift for tensorflow almost directly to support that style?